benchmarks/benchmark.py

```diff
@@ -0,0 +1,588 @@
+    Runs given command and waits until it finishes executing.
+    :param command: command to execute
+    """
+    print()
+    print(" ".join(command))
+    p1 = subprocess.Popen(command)
+    if p1.wait() != 0:
+        raise RuntimeError("command failed: " + " ".join(command))
+
+
+def build(exe_filepath):
+    """
+    Builds a file using make.
+    :param exe_filepath: File to build
+    """
+    run_command([make, exe_filepath])
+
+
+def run_benchmark(exe_filepath, n_repeats=1, csv_out_file=None):
+    """
+    Runs a benchmark
+    :param exe_filepath: path to the benchmark executable
+    :param n_repeats: how many times to repeat each benchmark
+    :param csv_out_file: path to csv fle to store benchmark results into
+    """
+    command = [exe_filepath]
+    if n_repeats > 1:
+        command.append("--benchmark_repetitions={}".format(n_repeats))
+        command.append("--benchmark_report_aggregates_only=true")
+    if csv_out_file is not None:
+        command.append("--benchmark_out={}".format(csv_out_file))
+        command.append("--benchmark_out_format=csv")
+    run_command(command)
+
+
+def pick_color(n):
+    str_bit_reversed_n = '{:015b}'.format(n + 1)[::-1]
+    r = 0.9 * ((int(str_bit_reversed_n[0::3], 2) / 2. ** 5 + 0.3) % 1)
+    g = 0.9 * ((int(str_bit_reversed_n[1::3], 2) / 2. ** 5 + 0.3) % 1)
+    b = 0.9 * ((int(str_bit_reversed_n[2::3], 2) / 2. ** 5 + 0.3) % 1)
+    return r, g, b
+
+
+def plot_results(csv_filename, out_file="", plot_log_y=False):
+    """
+    Plots benchmark results.
+    :param csv_filename: path to csv file containing results to plot
+    :param out_file: path to image file to store figure into. If it equals to "window" opens it in an interactive window.
+    """
+    import pandas
+    import numpy
+    import matplotlib
+    if out_file != "window":
+        matplotlib.use('Agg')
+    import matplotlib.pyplot as plt
+    with open(csv_filename) as f:
+        # google benchmark writes some non-csv data at beginning
+        for line in iter(f.readline, ''):
+            if line.startswith("name,iterations"):
+                f.seek(f.tell() - len(line) - 1, os.SEEK_SET)
+                break
+        data = pandas.read_csv(f)
+
+    timing_data = pandas.concat([data["name"].str.split("/", expand=True).iloc[:, :2], data["real_time"]], axis=1)
+    timing_data.columns = ["signatures", "sizes", "times"]
+    timing_data.loc[:, "sizes"] = timing_data["sizes"].astype(int)
+    timing_data.loc[:, "times"] /= 1000  # convert to microseconds
+
+    fig, ax = plt.subplots(figsize=(10, 10))
+    fig.set_tight_layout(True)
+    ax.set_xscale("log")
+    if plot_log_y:
+        ax.set_yscale("log")
+
+    ax.set_xlabel("size")
+    ax.set_ylabel("time[us]")
+    for n, (signature, sub_data) in enumerate(timing_data.groupby("signatures")):
+        # conversions to numpy arrays allow us to hide duplicated labels
+        ax.plot(numpy.array(sub_data["sizes"]), numpy.array(sub_data["times"]), "x",
```

Comment:
```suggestion
        ax.plot(sub_data["sizes"].values, sub_data["times"].values, "x",
```

Does this really create a duplicated labels with pandas?

Reply:
It seems matplotlib gets column labels out of data frames and sets them as labels. As I do two plots separately (lines and x-es) I get duplicated labels. I tried setting label=None, but it does not help.

Reply:
I think `label=__nolegend__`

Reply:
That works.